### PR TITLE
JSON RPC Engine

### LINKED ISSUE #1209: Contract events not getting through
I am using the redux store in `DrizzleProvider` to listen to contract events. On Desktop and on mobile Firefox (Android) with MetaMask as Add-on this works fine. But in MetaMask mobile, not a single event reaches the store. I connected via chrome-inspect to see the `console.log()` and there was nothing, not even errors. Therefore I assume you are either accidentally blocking contract events completely, or something in truffle/drizzle doesn't work with the way you do pass contract events...

Possibly same cause as #1111.

### LINKED ISSUE #1216: support EIP1102 `eth_requestAccounts`
currently, it doesn't work.

`ethereum.send('eth_requestAccounts', [])`

`ethereum.enable()` is deprecated

https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1102.md

image

### LINKED ISSUE #1232: autoRefreshOnNetworkChange not supported on metamask mobile
Setting 
```JS
ethereum.autoRefreshOnNetworkChange = false;
```
Doesn't make a difference on metamask mobile. It still reloads the page after a network change, unlink the metamask extension.
Also see: https://metamask.github.io/metamask-docs/API_Reference/Ethereum_Provider
(although this function will be removed in the future, it allows to prepare for the future updates)

### PR DESCRIPTION
**Pull request by brunobar79**
Replaces the whole provider implementation with https://github.com/metamask/metamask-mobile-provider which internally uses `metamask-inpage-provider` so we have no more differences between the extension provider and mobile

Fixes #1216  ( Parity with the extension)
Fixes #1209 
Fixes #1232

### CODE REVIEW

File: app/components/Views/BrowserTab/index.js [hunk 1]
```diff
@@ -53,12 +53,16 @@ import Branch from 'react-native-branch';
 import WatchAssetRequest from '../../UI/WatchAssetRequest';
 import Analytics from '../../../core/Analytics';
 import ANALYTICS_EVENT_OPTS from '../../../util/analytics';
-import { resemblesAddress } from '../../../util/address';
+//import { resemblesAddress } from '../../../util/address';
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: commented

File: app/components/Views/BrowserTab/index.js [hunk 2]
```diff
@@ -447,297 +450,278 @@ export class BrowserTab extends PureComponent {
...
+							if (req.method !== 'eth_requestAccounts') return next();
+							const { hostname } = senderUrl;
+							const { params } = req;
+							const { approvedHosts, privacyMode, selectedAddress } = this.props;
+							if (!privacyMode || ((!params || !params.force) && approvedHosts[hostname])) {
+								res.result = [selectedAddress];
+							} else {
+								if (!this.state.showApprovalDialog) {
+									setTimeout(async () => {
+										if (!this.state.showApprovalDialog) {
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: is checking the same two lines above

File: app/components/Views/BrowserTab/index.js [hunk 3]
```diff
@@ -447,297 +450,278 @@ export class BrowserTab extends PureComponent {
...
+											await this.getPageMeta();
+											this.setState({
+												showApprovalDialog: true,
+												showApprovalDialogHostname: hostname
+											});
+										}
+									}, 1000);
+								}
+								const approved = await new Promise((resolve, reject) => {
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: this doesn't crash when `approvalRequest.reject('User rejected account access')`?

Comment (estebanmino) [reply app/components/Views/BrowserTab/index.js]: hm I see `createAsyncMiddleware` handles it

File: app/components/Views/BrowserTab/index.js [hunk 4]
```diff
@@ -447,297 +450,278 @@ export class BrowserTab extends PureComponent {
...
+					eth_signTypedData_v3: () =>
+						createAsyncMiddleware(async (req, res, next) => {
+							if (req.method !== 'eth_signTypedData_v3') return next();
+							const { TypedMessageManager } = Engine.context;
+							const data = JSON.parse(req.params[1]);
+							const chainId = data.domain.chainId;
+							const activeChainId =
+								this.props.networkType === 'rpc'
+									? this.props.network
+									: Networks[this.props.networkType].networkId;
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: I've been using `this.props.network` here, you get network id from the controller

File: app/components/Views/BrowserTab/index.js [hunk 5]
```diff
@@ -447,297 +450,278 @@ export class BrowserTab extends PureComponent {
...
+
+							// eslint-disable-next-line eqeqeq
+							if (chainId && chainId != activeChainId) {
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: chainId !=`=` activeChainId
and I believe you can remove eslint comment

File: app/components/Views/BrowserTab/index.js [hunk 6]
```diff
@@ -447,297 +450,278 @@ export class BrowserTab extends PureComponent {
...
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: commented the same above

File: app/components/Views/BrowserTab/index.js [hunk 7]
```diff
@@ -447,297 +450,278 @@ export class BrowserTab extends PureComponent {
...
+									data: req.params[1],
+									from: req.params[0],
+									...pageMeta
+								},
+								'V4'
+							);
+
+							res.result = rawSig;
+						}),
+					web3_clientVersion: () =>
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: I'm missing `net_version`call

File: app/core/BackgroundBridge.js [hunk 8]
```diff
@@ -1,115 +1,241 @@
...
+		engine.push(this.middlewares.eth_signTypedData_v3());
+		engine.push(this.middlewares.eth_signTypedData_v4());
+
+		// walletMethods
+		engine.push(this.middlewares.web3_clientVersion());
+		engine.push(this.middlewares.wallet_scanQRCode());
+		engine.push(this.middlewares.wallet_watchAsset());
+
+		// Mobile specific methods
+		engine.push(this.middlewares.wallet_watchAsset());
```
Comment (estebanmino) [inline app/core/BackgroundBridge.js]: https://github.com/MetaMask/metamask-mobile/pull/1236/files#diff-119ae1f9db5d86a529bfd0fbabbfab91R126 ?

File: app/core/BackgroundBridge.js [hunk 9]
```diff
@@ -1,115 +1,241 @@
...
-		if (!this._accounts) {
-			this._accounts = true;
-			this.sendStateUpdate();
-		}
+	getState() {
+		const vault = Engine.context.KeyringController.state.vault;
+		const isInitialized = !!vault;
+		const { network, selectedAddress } = Engine.datamodel.flatState;
+		return {
+			...{ isInitialized },
```
Comment (estebanmino) [inline app/core/BackgroundBridge.js]: just `isInitialized,` ?

File: app/core/BackgroundBridge.js [hunk 10]
```diff
@@ -1,115 +1,241 @@
...
+			const isEnabled = checkIsEnabled();
+			const isReady = isUnlocked && isEnabled;
+			const networkType = Engine.context.NetworkController.state.provider.type;
+			const chainId = Object.keys(NetworkList).indexOf(networkType) > -1 && NetworkList[networkType].chainId;
+			const result = {
+				isUnlocked,
+				isEnabled,
+				selectedAddress: isReady ? selectedAddress.toLowerCase() : null,
+				networkVersion: network,
+				chainId: `0x${parseInt(chainId, 10).toString(16)}`
```
Comment (estebanmino) [inline app/core/BackgroundBridge.js]: This won't crash the app but I for rpc networks we'll be returning `0xNaN` if that's ok 👍

File: app/components/Views/BrowserTab/index.js (hunk 2 above)
Comment (brunobar79) [reply app/components/Views/BrowserTab/index.js]: The check is the same but the second one is done later in time.  That handles the scenario when an iframe and the parent frame is requesting account access at the same time, preventing two modals to show up.

File: app/components/Views/BrowserTab/index.js (hunk 4 above)
Comment (brunobar79) [reply app/components/Views/BrowserTab/index.js]: This is the same exact implementation that we have right now in prod: https://github.com/MetaMask/metamask-mobile/blob/develop/app/components/Views/BrowserTab/index.js#L535

File: app/components/Views/BrowserTab/index.js (hunk 5 above)
Comment (brunobar79) [reply app/components/Views/BrowserTab/index.js]: Same here: https://github.com/MetaMask/metamask-mobile/blob/develop/app/components/Views/BrowserTab/index.js#L539

To use === I'll have to do `.toString().toLowerCase()` on both ends.

File: app/components/Views/BrowserTab/index.js (hunk 6 above)
Comment (brunobar79) [reply app/components/Views/BrowserTab/index.js]: Same reply as above

File: app/components/Views/BrowserTab/index.js (hunk 7 above)
Comment (brunobar79) [reply app/components/Views/BrowserTab/index.js]: Handled by `metamask-inpage-provider` https://github.com/MetaMask/metamask-inpage-provider/blob/master/index.js#L346 . No need to go to the background for info that the page already has

File: app/core/BackgroundBridge.js (hunk 8 above)
Comment (brunobar79) [reply app/core/BackgroundBridge.js]: Removed the comment

File: app/core/BackgroundBridge.js (hunk 10 above)
Comment (brunobar79) [reply app/core/BackgroundBridge.js]: added a check and set it to null

**Comment by ibrahimtaveras00** [comment]
**Issue 1:**

The tutorial seems to have an error, only on Android when you go through the flow all the way to browser, then go back a few times when you're back on wallet

When creating a new wallet flow: seen here = http://recordit.co/Yo23vkYbmP

image

When Synced already via extension and you tap on the "Take a tour" box at the bottom of our homepage in browser; seen here = https://recordit.co/22W3c8Y1rt

Screen Shot 2019-12-06 at 4 27 33 PM

**Issue 2:** 

On playstore/testflight when I go to uniswap and connect, pretty quickly I can see my balance on top of the ETH dropdown, however, on this branch, I don't see my balance until I reload the page

seen here =  http://recordit.co/7ZDqImuJ5Q

Happens on both android + iOS

**Issue 3:**

Happening on both android + iOS; if I go to migrate.makerdao.com and connect, the next screen just shows an endless spinner = https://recordit.co/JAeGLqcZIf

here's comparison to testflight = http://recordit.co/7ZDqImuJ5Q 

**Issue 4:**

I'm having issues getting the connect modal to pop up when navigating to https://3box.io/ and tapping on choose wallet or log in.  For android, I have to reload the page, then it works; while on iOS reloading the page does nothing.

Seen here = http://recordit.co/Q3tLcbgzBv

**Comment by ibrahimtaveras00** [comment]
Good news, I just built latest develop and can knock some things off this list 

Issue 1 happens on develop but not playstore (I'll create a different issue for this) 

Issue 6 same, but this time I can reproduce on playstore as well compared to last week (I'll create a different issue for this)

Issue 7 Same

**Comment by brunobar79** [comment]
@ibrahimtaveras00  just pushed all the fixes.

@estebanmino  Could you please take another look ?

Issue 1 - Uniswap - FIXED
Issue 2 - migrate.makerdao.com - FIXED
Issue 3 - 3box - FIXED
Issue 4 - That error is fixed, however there's a problem on peeepeth side that will fail in the very last step saying "Error Unprocessable entity" which is an issue with peepeth itself.
Issue 5 - Aragon - FIXED
Issue 6 - Wyre - FIXED
Issue 7 -  ETHFinxes - FIXED
Issue 8 - KyberSwap - FIXED

File: app/components/Views/BrowserTab/index.js [hunk 11]
```diff
@@ -447,297 +449,273 @@ export class BrowserTab extends PureComponent {
...
+					}),
+
+				eth_accounts: senderUrl =>
+					createAsyncMiddleware(async (req, res, next) => {
+						if (req.method !== 'eth_accounts') return next();
+						const { hostname } = senderUrl;
+						const { approvedHosts, privacyMode, selectedAddress } = this.props;
+						const isEnabled = !privacyMode || approvedHosts[hostname];
+						if (isEnabled) {
+							res.result = [selectedAddress.toLowerCase()];
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js]: why `toLowerCase`? we have it on checksummed format

**Review by estebanmino — APPROVED** [review]
LGTM 👍

File: app/components/Views/BrowserTab/index.js (hunk 11 above)
Comment (brunobar79) [reply app/components/Views/BrowserTab/index.js]: Better safe than sorry

**Review by ibrahimtaveras00 — APPROVED** [review]
looks good now on both OS's, QA Passed 👍

**Pull request merged**